src/backend/auth/azure.rs

```diff
@@ -0,0 +1,266 @@
+            if_match = headers(header::IF_MATCH)
+                .map(|v| v.to_str().expect("if-match should be a string"))
+                .unwrap_or(""),
+            if_none_match = headers(header::IF_NONE_MATCH)
+                .map(|v| v.to_str().expect("if-none-match should be a string"))
+                .unwrap_or(""),
+            if_unmodified_since = headers(header::IF_UNMODIFIED_SINCE)
+                .map(|v| v.to_str().expect("if-unmodified-since should be a string"))
+                .unwrap_or(""),
+            range = headers(header::RANGE)
+                .map(|v| v.to_str().expect("range should be a string"))
+                .unwrap_or(""),
+        )
+    }
+
+    /// Formats a canonical header string given the Microsoft extension headers.
+    ///
+    /// See: <https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key#constructing-the-canonicalized-headers-string>
+    ///
+    /// # Panics
+    ///
+    /// Panics if any of the header values cannot be represented as a string or
+    /// if there is a duplicate header value.
+    fn canonical_headers<'b>(
+        &self,
+        microsoft_headers: impl Iterator<Item = (&'b HeaderName, &'b HeaderValue)>,
+    ) -> String {
+        let mut headers = BTreeMap::new();
+        for (k, v) in microsoft_headers {
+            if headers
+                .insert(
+                    k.as_str(),
+                    v.to_str().expect("expected a string value").trim(),
+                )
+                .is_some()
+            {
+                panic!("duplicate header `{k}`", k = k.as_str());
+            }
+        }
+
+        let mut canonical_headers = String::new();
+        for (k, v) in headers {
+            canonical_headers.push_str(k);
+            canonical_headers.push(':');
+            canonical_headers.push_str(v);
+            canonical_headers.push('\n');
+        }
+
+        canonical_headers
+    }
+
+    /// Formats a canonical resource string.
+    ///
+    /// See: https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key#constructing-the-canonicalized-resource-string
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is a duplicate query parameter.
+    fn canonical_resource<'b>(
+        &self,
+        path: &str,
+        query_pairs: impl Iterator<Item = (Cow<'b, str>, Cow<'b, str>)>,
+    ) -> String {
+        let mut canonical_resource = String::new();
+
+        canonical_resource.push('/');
+        canonical_resource.push_str(self.0.account_name());
+        canonical_resource.push_str(path);
+
+        let mut parameters = BTreeMap::new();
+        for (k, v) in query_pairs {
+            if parameters.insert(k.clone(), v).is_some() {
+                panic!("duplicate query parameter `{k}`");
+            }
+        }
+
+        for (k, v) in parameters {
+            canonical_resource.push('\n');
+            canonical_resource.push_str(&k);
```

Comment:
The specification states that query parameter names should be converted to lowercase.

```suggestion
            canonical_resource.push_str(&k.to_lowercase());
```

Reply:
I was under the mistaken impression that the parse would normalize it, but it doesn't.

I'll fix.